### PR TITLE
feat(clone): add support for creating the Clone from volume as datasource

### PR DESCRIPTION
**Pull request by pawanpraka1**
**Why is this PR required? What issue does it fix?**:

 add support for creating the Clone from volume as datasource

**What this PR does?**:

We can now create the Clone from pvc directly

```
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: pvc-clone
spec:
  storageClassName: openebs-clone
  dataSource:
    name: pvc-vol
    kind: PersistentVolumeClaim
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 4Gi 
```
The ZFS_LocalPV driver will create one internal snapshot of the name
same as the new volume name and will create a clone out of it. Also,
while destroying the volume the driver will take care of deleting
the created snapshot for the clone.

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated?
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track:
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

### CODE REVIEW

**Comment by kmova** [comment]
@pawanpraka1 - what happens in the following case: 

- create pvc1
- create pvc2 ( clone of pvc1)
- delete pvc1 

Will pvc2 (and its PV) still exist?

**Comment by pawanpraka1** [comment]
Yes @kmova, pvc2 and it's pv will be there. Also zfs volume created for pvc1 will also be there and it will be cleaned up automatically when we delete pvc2.

File: pkg/driver/controller.go [hunk 1]
```diff
@@ -165,8 +165,56 @@ func CreateZFSVolume(req *csi.CreateVolumeRequest) (string, error) {
...
+			vol.Spec.PoolName, pool)
+	}
+
+	if vol.Spec.Capacity != volsize {
+		return "", status.Error(codes.Internal, "clone: volume size is not matching")
+	}
+
+	selected := vol.Spec.OwnerNodeID
+
+	labels := map[string]string{zfs.ZFSVolKey: vol.Name}
```
Comment (kmova) [inline pkg/driver/controller.go:194]: Can we make this a more explicit key - like "source-zfs-volume"?

File: pkg/zfs/zfs_util.go [hunk 2]
```diff
@@ -580,6 +600,27 @@ func DestroyVolume(vol *apis.ZFSVolume) error {
...
+		// datasource is volume, delete the dependent snapshot
+		snap := &apis.ZFSSnapshot{}
+		snap.Name = vol.Name // snapname is same as volname
+		snap.Spec = vol.Spec
+		// add src vol name
+		snap.Labels = map[string]string{ZFSVolKey: srcVol}
+
+		klog.Infof("destroying snapshot %s@%s for the clone %s", srcVol, snap.Name, volume)
+
+		err := DestroySnapshot(snap)
```
Comment (kmova) [inline pkg/zfs/zfs_util.go:614]: Here is where the source volume needs to be cleaned up, in case this was the last clone/snapshot being deleted?

Comment (pawanpraka1) [reply pkg/zfs/zfs_util.go:614]: no, not for the last clone/snapshot being deleted, every clone will have a corresponsing snapshot. So when we are deleting the clone volume, we should delete the corresponding snapshot here.

**Review by kmova — APPROVED** [review]
this PR will be followed up documentation updates and better log management in case of error cases.

**Comment by pawanpraka1** [comment]
@kmova here is the issue https://github.com/openebs/zfs-localpv/issues/236

**Pull request merged**